### PR TITLE
just some minor tweaks

### PR DESCRIPTION
**Pull request by cirex-web**
I was just looking through the typedefs to generate interfaces for CMUEats and I couldn't help but make some small changes here and there... 

Among these changes, I
- made `ILocation` a bit more strict than before
- Removed some type assertions that may not be true
- Removed "redundant" code (I can rollback anything that seems sus)
- Added retry limit of 4 to `getHTMLResponse` (just in case bad network connection caused a request to fail)
- Moved location open times sort function from `server.ts` to the build function in `timeBuilder.ts` (separation of concerns I guess)

### CODE REVIEW

**Comment by GhostOf0days** [comment]
Thank you for the PR. Due to my workload this week, I'll review this the upcoming weekend. I apologize for the delay.

**Comment by cirex-web** [comment]
Hmm good point, but wouldn't having looser restrictions also mean that the API might be unreliable and run the risk of returning undefined values for potentially any attribute? Would we then have to validate the data in the CMU Eats code instead of here?

**Comment by GhostOf0days** [comment]
Those cases are very rare based on previous iterations of the dining website and how we fixed them, and would probably take more time to fix with stricter restrictions. I would say yes to validating data on CMUEats instead of the dining API.

**Comment by cirex-web** [comment]
all fields?

**Comment by cirex-web** [comment]
Also, is failing fast on the API side really a problem? (since we could just use the cached data indefinitely)

**Comment by GhostOf0days** [comment]
I would probably say it's still a problem. As for the stricter internal API, I'm slightly confused. Could you elaborate? If you suggest we have a separate API that's stricter, the issue lies with it not being worth the extra cost for build hours.

File: src/containers/locationBuilder.ts [hunk 1]
```diff
@@ -114,8 +115,20 @@ export default class LocationBuilder {
...
+  }
+  isValid() {
+    return this.valid;
+  }
   build(): ILocation {
+    if (!this.valid) throw Error("Location has been invalidated!");
+    if (this.times === undefined || this.acceptsOnlineOrders === undefined || this.description === undefined
+      || this.url === undefined || this.location === undefined) {
+      throw Error("Didn't finish configuring restaurant before building metadata!");
+      //all fetches were good - yet we have missing data. This is a problem
```
Comment (cirex-web) [inline src/containers/locationBuilder.ts:129]: This was the extent of my _stricter_ validation - sorry if there was any confusion.

**Comment by cirex-web** [comment]
I defined a stricter API as just code with more fields are undefined/null-checked before being made available to the client.
https://github.com/ScottyLabs/dining-api/pull/111/files#r1428176263

**Comment by cirex-web** [comment]
While working on CMU Eats today I ran into an assertion error in one of my sanity check functions. It turns out that the opening times were being sorted improperly in the API. (This is important for `getNextTimeSlot`)

Pasted Graphic 5


Turns out, the sorting function in `server.ts` returns -1 in places where it’s supposed to return 1. This results in improperly sorted times (see img 2)

Screenshot 2023-12-22 at 3 54 31 PM


Pasted Graphic 3

Being sorted is a key assumption of `getNextTimeSlot` (in both the new and old code), so an unsorted times array leads to wrong opening times occasionally. For instance, rewinding back to 7AM on Friday (day: 5) gets us
Pasted Graphic 6
, which is wrong. (should say opening at 10AM)


Inadvertently, in cleaning up the sorting function last week by replacing it with the following, I also removed the bug.

Pasted Graphic 4
Pasted Graphic 2
(times are sorted, as expected)

I also made a separate emergency PR that just fixes the sorting problem - but if you get the chance we can merge this one in to also fix the problem.

I suppose the lesson is that using `assert` statements is a good way to find stuble bugs.

**Comment by GhostOf0days** [comment]
Let's revert changes in TimeUtils file and change TWENTYFOUR_HOURS back to TWENTYFOURHOURS then I'll re-review.

**Comment by GhostOf0days** [comment]
Let's revert changes in timeUtils file and change TWENTYFOUR_HOURS back to TWENTYFOURHOURS then I'll re-review.

**Comment by GhostOf0days** [comment]
Also, let's make location optional.

**Comment by cirex-web** [comment]
> Also, let's make description optional.

Description will always be defined as $("").text() will return a string even if the element isn't found.

Relevant code snippet:
image


Testing cheerio library:
image

**Comment by GhostOf0days** [comment]
Try now.

File: src/parser/diningParser.ts [hunk 2]
```diff
@@ -145,33 +151,39 @@ export default class DiningParser {
...
-  
-      if (Array.isArray(specialList)) {
-        builder.setSpecials(specialList);
-      }
-  
-      if (Array.isArray(soupList)) {
-        builder.setSoups(soupList);
+      const name = builder.getName();
+      if (name !== undefined) {
+        //Not sure when name would be undefined, but just in case...
```
Comment (GhostOf0days) [inline src/parser/diningParser.ts:164]: Remove comment here

**Review by GhostOf0days — COMMENTED** [review]
Great. Passes all my testing, and all code changes look good to me. One small thing. Remove the comment I marked, run ```ncu -u```, then run ```npm i```. I'll merge the PR after.

**Comment by cirex-web** [comment]
wait hang on don't merge yet - I want to see what's causing the package-lock conflict

**Comment by cirex-web** [comment]
also made a few minor tweaks that I squashed into this commit https://github.com/ScottyLabs/dining-api/pull/111/commits/60b888bd3e9e550cdd7cbd7699007f1aae806396

**Comment by GhostOf0days** [comment]
Alright. I'll let you fix it then for learning purposes. I already know what's causing the merge conflicts.

**Comment by GhostOf0days** [comment]
Also, yeah. Those changes make sense.

**Comment by cirex-web** [comment]
image

Ah so I bumped @types/node from 20.10.7 to 20.11.0 but dependabot bumped it to 20.10.8

**Comment by cirex-web** [comment]
lol actually yeah it was kinda obvious
image

**Comment by GhostOf0days** [comment]
lmao

**Comment by GhostOf0days** [comment]
i can merge now?

**Comment by cirex-web** [comment]
wait I'm so confused lol I literally cherry-picked the dependabot commit and then ran ncu on that commit

**Comment by cirex-web** [comment]
oh god this is what a rebase looks like

**Comment by GhostOf0days** [comment]
💀

**Comment by cirex-web** [comment]
shoot I thought origin was this repo
image

**Comment by GhostOf0days** [comment]
Let’s uhh do something really quick

**Comment by cirex-web** [comment]
honestly yeah we could just _fix the merge conflict_ but I was thinking of doing it a different way

**Comment by GhostOf0days** [comment]
Run ```git log``` then ```git reset --hard commit-id-before-rebase``` or ```git reset --hard HEAD~n```, where n is the number of commits we want to go back by. Then push that.

**Comment by GhostOf0days** [comment]
I once had merge conflicts in 7 or so different files in CMUEats. I didn’t have to push anything to main, but I don’t remember how I fixed it. I would have to check. I remember I did use some sort of rebase.

**Comment by cirex-web** [comment]
I forgot to pull main 💀

**Comment by cirex-web** [comment]
legit rebased on an outdated version

**Comment by GhostOf0days** [comment]
Lmao

**Comment by GhostOf0days** [comment]
See if that hard reset before rebase then push works. I sure hope so. 💀

**Comment by cirex-web** [comment]
> 
Run git log then git reset --hard commit-id-before-rebase or git reset --hard HEAD~n, where n is the number of commits we want to go back by. Then push that.

I have a feeling that's just going to delete everything here cuz I force pushed the rebase (the old branch is basically gone)

**Comment by cirex-web** [comment]
but I'll try

**Comment by GhostOf0days** [comment]
If it doesn’t work, let’s uhh. Idk.

**Comment by cirex-web** [comment]
☠️ ☠️ ☠️

**Pull request closed without merging**